provider/apple.go

```diff
@@ -0,0 +1,534 @@
+	if _, _, err := ah.JwtService.Get(r); err != nil {
+		rest.SendErrorJSON(w, r, ah.L, http.StatusForbidden, err, "logout not allowed")
+		return
+	}
+	ah.JwtService.Reset(w)
+}
+
+// exchange sends the validation token request and gets access token and user claims
+// (e.g. https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens)
+func (ah *AppleHandler) exchange(ctx context.Context, code, redirectURI string, result interface{}) (err error) {
+
+	// check client_secret is expired and recreate new (client_secret JWT) if need
+	ok, err := ah.isClientSecretExpired()
+	if err != nil || !ok {
+		ah.conf.clientSecret, err = ah.createClientSecret()
+		if err != nil {
+			return errors.Wrap(err, "client secret create failed")
+		}
+	}
+
+	data := url.Values{}
+	data.Set("client_id", ah.conf.ClientID)
+	data.Set("client_secret", ah.conf.clientSecret) // JWT signed with Apple private key
+	data.Set("code", code)
+	data.Set("redirect_uri", redirectURI) // redirect URL can't refer to localhost and must have trusted certificate and https protocol
+	data.Set("grant_type", "authorization_code")
+
+	client := http.Client{Timeout: time.Second * 5}
+	req, err := http.NewRequestWithContext(ctx, "POST", ah.endpoint.TokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("content-type", appleRequestContentType)
+	req.Header.Add("accept", AcceptHeader)
+	req.Header.Add("user-agent", ah.conf.UserAgent) // apple requires a user agent
+
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	// Apple REST API response either 200 (ok OK) or 400 (if error)
+	if res.StatusCode >= 400 {
+		return fmt.Errorf("exchange response error code: %d", res.StatusCode)
+	}
+
+	err = json.NewDecoder(res.Body).Decode(result)
+	defer func() {
+		if err = res.Body.Close(); err != nil {
+			ah.L.Logf("[ERROR] close request body failed when get access token: %v", err)
+		}
+	}()
+	return err
+}
+
+// getTokenClaims decodes the id_token in response for extract claims with user data
+func (ah AppleHandler) getTokenClaims(idToken string) (jwt.MapClaims, error) {
+	tkn, _, err := new(jwt.Parser).ParseUnverified(idToken, jwt.MapClaims{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	claims, ok := tkn.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, errors.New("can't convert token claims to standard claims")
+	}
+	return claims, nil
+}
+
+// createClientSecret use for create the JWT client secret required to make requests to the Apple validation server.
+// for more details go to link: https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens#3262048
+func (ah *AppleHandler) createClientSecret() (string, error) {
+	if ah.conf.privateKey == nil {
+		return "", errors.New("private key can't be empty")
+	}
+	// Create the Claims
+	now := time.Now()
+	exp := now.Add(time.Second * 86400).Unix() // default value
```

Comment:
is it 24 hours? do we really need it that long?

Reply:
No. What are you offer?  1 hour?

Reply:
maybe even shorter? for the verified provider we do 30m

Reply:
ok